### ISSUE TITLE
where is "self.getGradCam"

### ISSUE THREAD

**Issue by Zhou-AY** [whdii/TMM#3]
In line 82 of file multimodalAttack.py, where is "self.getGradCam" defined? Could you please improve the code？Thank you.

**Comment by chilljudaoren** [comment]
请问你的问题解决了嘛？你觉得他这个代码能跑通嘛？

**Comment by chilljudaoren** [comment]
没有self.getGradCam部分呀?

**Comment by chilljudaoren** [comment]
def getGradCam(self, gradcam, shape):
        # 示例方法，假设对 gradcam 进行插值以匹配给定的形状
        gradcam = torch.nn.functional.interpolate(gradcam.unsqueeze(0), size=shape, mode='bilinear', align_corners=False)
        return gradcam.squeeze(0)

**Comment by chilljudaoren** [comment]
请问您解决了嘛？

**Comment by NingpingMou** [comment]
> ```
> def getGradCam(self, gradcam, shape):
>     # 示例方法，假设对 gradcam 进行插值以匹配给定的形状
>     gradcam = torch.nn.functional.interpolate(gradcam.unsqueeze(0), size=shape, mode='bilinear', align_corners=False)
>     return gradcam.squeeze(0)
> ```

请问您给出的定义，是作者实现的吗

**Comment by chilljudaoren** [comment]
这个我是问gpt的

NingpingMou ***@***.***> 于2024年10月15日周二 16:58写道：

**Comment by YXL2240901958** [comment]
> 请问您解决了嘛？

请问能跟您交流一下吗 id是我qq

**Comment by chilljudaoren** [comment]
> > 请问您解决了嘛？
> 
> 请问能跟您交流一下吗 id是我qq
可以呀

**Comment by chilljudaoren** [comment]
def compute_gradcam(model, visual_input, text_input, tokenized_text, block_num=6):
    model.text_encoder.base_model.base_model.encoder.layer[
        block_num
    ].crossattention.self.save_attention = True

    output = model({"image": visual_input, "text_input": text_input}, match_head="itm")
    loss = output[:, 1].sum()

    model.zero_grad()
    loss.backward()
    with torch.no_grad():
        mask = tokenized_text.attention_mask.view(
            tokenized_text.attention_mask.size(0), 1, -1, 1, 1
        )  # (bsz,1,token_len, 1,1)
        token_length = tokenized_text.attention_mask.sum(dim=-1) - 2
        token_length = token_length.cpu()
        # grads and cams [bsz, num_head, seq_len, image_patch]
        grads = model.text_encoder.base_model.base_model.encoder.layer[
            block_num
        ].crossattention.self.get_attn_gradients()
        cams = model.text_encoder.base_model.base_model.encoder.layer[
            block_num
        ].crossattention.self.get_attention_map()

        # assume using vit with 576 num image patch
        cams = cams[:, :, :, 1:].reshape(visual_input.size(0), 12, -1, 24, 24) * mask
        grads = (
            grads[:, :, :, 1:].clamp(0).reshape(visual_input.size(0), 12, -1, 24, 24)
            * mask
        )

        gradcams = cams * grads
        gradcam_list = []

        for ind in range(visual_input.size(0)):
            token_length_ = token_length[ind]
            gradcam = gradcams[ind].mean(0).cpu().detach()
            # [enc token gradcam, average gradcam across token, gradcam for individual token]
            gradcam = torch.cat(
                (
                    gradcam[0:1, :],
                    gradcam[1 : token_length_ + 1, :].sum(dim=0, keepdim=True)
                    / token_length_,
                    gradcam[1:, :],
                )
            )
            gradcam_list.append(gradcam)
            
    return gradcam_list, output

**Comment by LeeZSir** [comment]
> def compute_gradcam(model, visual_input, text_input, tokenized_text, block_num=6): model.text_encoder.base_model.base_model.encoder.layer[ block_num ].crossattention.self.save_attention = True
> 
> ```
> output = model({"image": visual_input, "text_input": text_input}, match_head="itm")
> loss = output[:, 1].sum()
> 
> model.zero_grad()
> loss.backward()
> with torch.no_grad():
>     mask = tokenized_text.attention_mask.view(
>         tokenized_text.attention_mask.size(0), 1, -1, 1, 1
>     )  # (bsz,1,token_len, 1,1)
>     token_length = tokenized_text.attention_mask.sum(dim=-1) - 2
>     token_length = token_length.cpu()
>     # grads and cams [bsz, num_head, seq_len, image_patch]
>     grads = model.text_encoder.base_model.base_model.encoder.layer[
>         block_num
>     ].crossattention.self.get_attn_gradients()
>     cams = model.text_encoder.base_model.base_model.encoder.layer[
>         block_num
>     ].crossattention.self.get_attention_map()
> 
>     # assume using vit with 576 num image patch
>     cams = cams[:, :, :, 1:].reshape(visual_input.size(0), 12, -1, 24, 24) * mask
>     grads = (
>         grads[:, :, :, 1:].clamp(0).reshape(visual_input.size(0), 12, -1, 24, 24)
>         * mask
>     )
> 
>     gradcams = cams * grads
>     gradcam_list = []
> 
>     for ind in range(visual_input.size(0)):
>         token_length_ = token_length[ind]
>         gradcam = gradcams[ind].mean(0).cpu().detach()
>         # [enc token gradcam, average gradcam across token, gradcam for individual token]
>         gradcam = torch.cat(
>             (
>                 gradcam[0:1, :],
>                 gradcam[1 : token_length_ + 1, :].sum(dim=0, keepdim=True)
>                 / token_length_,
>                 gradcam[1:, :],
>             )
>         )
>         gradcam_list.append(gradcam)
>         
> return gradcam_list, output
> ```

您好，能加一个联系方式进一步交流嘛？

**Comment by YXL2240901958** [comment]
> > def compute_gradcam(model, visual_input, text_input, tokenized_text, block_num=6): model.text_encoder.base_model.base_model.encoder.layer[ block_num ].crossattention.self.save_attention = True
> > ```
> > output = model({"image": visual_input, "text_input": text_input}, match_head="itm")
> > loss = output[:, 1].sum()
> > 
> > model.zero_grad()
> > loss.backward()
> > with torch.no_grad():
> >     mask = tokenized_text.attention_mask.view(
> >         tokenized_text.attention_mask.size(0), 1, -1, 1, 1
> >     )  # (bsz,1,token_len, 1,1)
> >     token_length = tokenized_text.attention_mask.sum(dim=-1) - 2
> >     token_length = token_length.cpu()
> >     # grads and cams [bsz, num_head, seq_len, image_patch]
> >     grads = model.text_encoder.base_model.base_model.encoder.layer[
> >         block_num
> >     ].crossattention.self.get_attn_gradients()
> >     cams = model.text_encoder.base_model.base_model.encoder.layer[
> >         block_num
> >     ].crossattention.self.get_attention_map()
> > 
> >     # assume using vit with 576 num image patch
> >     cams = cams[:, :, :, 1:].reshape(visual_input.size(0), 12, -1, 24, 24) * mask
> >     grads = (
> >         grads[:, :, :, 1:].clamp(0).reshape(visual_input.size(0), 12, -1, 24, 24)
> >         * mask
> >     )
> > 
> >     gradcams = cams * grads
> >     gradcam_list = []
> > 
> >     for ind in range(visual_input.size(0)):
> >         token_length_ = token_length[ind]
> >         gradcam = gradcams[ind].mean(0).cpu().detach()
> >         # [enc token gradcam, average gradcam across token, gradcam for individual token]
> >         gradcam = torch.cat(
> >             (
> >                 gradcam[0:1, :],
> >                 gradcam[1 : token_length_ + 1, :].sum(dim=0, keepdim=True)
> >                 / token_length_,
> >                 gradcam[1:, :],
> >             )
> >         )
> >         gradcam_list.append(gradcam)
> >         
> > return gradcam_list, output
> > ```
> 
> 您好，能加一个联系方式进一步交流嘛？

您好 这是我的联系方式 2240901958 qq

**Comment by chilljudaoren** [comment]
qq:905451466

LeeZSir ***@***.***> 于2025年1月2日周四 10:58写道：
